orttraining/orttraining/models/bert/main.cc

```diff
@@ -379,6 +379,19 @@ Status ParseArguments(int argc, char* argv[], BertParameters& params, OrtParamet
     // Backward pass and optimizer nodes are implicitly generated by ORT.
     params.pipeline_stage_paths = flags["pipeline_stage_paths"].as<std::vector<std::string>>();
 
+    // If user doesn't provide partitioned model files, a cut list should be provided for ORT to do partition
+    // online. If the pipeline contains n stages, the cut list should be of length (n-1), in order to cut the
+    // graph into n partitions.
+    // TODO: find a better way (maybe a parser?) to better handle custom cut information.
+    if (params.pipeline_parallel_size > 1 && params.pipeline_stage_paths.empty()) {
+      TrainingSession::TrainingConfiguration::CutInfo cut0 = {TrainingSession::TrainingConfiguration::CutEdge("186"),
+                                                              TrainingSession::TrainingConfiguration::CutEdge("71", {"273"})};
+      TrainingSession::TrainingConfiguration::CutInfo cut1 = {TrainingSession::TrainingConfiguration::CutEdge("308"),
+                                                              TrainingSession::TrainingConfiguration::CutEdge("71", {"395"})};
+      params.pipeline_partition_cut_list.emplace_back(cut0);
```

Comment:
Those should be input arguments of main. For example, --cuts 186-71-273,308-71-395.

Reply:
yeah, preferably, it should come as flexible commandline input instead of c++ code. (see TODO on line 385)

Currently the parser is not capable of parsing structured data like this. (the examle you put is not sufficient as there are many ways to interpolate 186-71-273. But it shouldn't be hard. I'll add the parser in later.

---
In reply to: [429520899](https://github.com/microsoft/onnxruntime/pull/3996#discussion_r429520899) [](ancestors = 429520899)